perf_events_poller.go

```diff
@@ -0,0 +1,124 @@
+// Copyright (c) 2019 Dropbox, Inc.
+// Full license can be found in the LICENSE file.
+
+package goebpf
+
+/*
+#include <poll.h>
+#include <stdlib.h>
+#include <string.h>
+#include <stdio.h>
+
+static int perf_events_poll(void *_fds, int cnt, int timeout)
+{
+    int *fds = _fds;
+    int *fds_end = fds + cnt;
+
+    // Allocate N pollfd structs
+    size_t pollfds_size = sizeof(struct pollfd) * cnt;
+    void *pollfds_memory = malloc(pollfds_size);
+    memset(pollfds_memory, 0, pollfds_size);
+
+    // Initialize pollfds from GO array of uint32 fds
+    struct pollfd *pollfds = pollfds_memory;
+    for (; fds != fds_end; fds++, pollfds++) {
+        pollfds->fd = *fds;
+        pollfds->events = POLLIN;
+    }
+    // Re-set pointers to start of arrays
+    pollfds = pollfds_memory;
+    fds = _fds;
+
+    int ready_cnt = poll(pollfds, cnt, timeout);
+
+    // Copy all ready descriptors back into golang array of uint32s
+    for (int remain = ready_cnt; remain > 0; pollfds++) {
+        if (pollfds->revents & POLLIN) {
```

Comment:
from man:
" The bits returned in revents can
       include any of those specified in events, or one of the values
       POLLERR, POLLHUP, or POLLNVAL. "

so e.g. if there was any event which is not POLLIN we would never have remain == 0 and would try to access thinigs outside pollfd's allocated memory (w/ eventual segfolt)

Reply:
Good catch, added one more condition to avoid this:
```
pollfds != pollfds_end
```